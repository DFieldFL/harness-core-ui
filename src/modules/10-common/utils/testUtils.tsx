import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uikit'
import qs from 'qs'

import strings from 'strings/strings.en.yaml'
import { AppStoreProvider, useAppStoreWriter } from 'framework/hooks/useAppStore'
import type { AppStore } from 'framework/types/AppStore'
import { AUTH_ROUTE_PATH_PREFIX, StringsContext } from 'framework/exports'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): Element | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): Element | null => document.querySelector('.bp3-popover-content')

interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, any>
  defaultAppStoreValues?: Partial<AppStore>
}

export const prependAccountPath = (path: string): string => AUTH_ROUTE_PATH_PREFIX + path

const AppStoreInitializer: React.FC<{ defaultAppStoreValues: TestWrapperProps['defaultAppStoreValues'] }> = ({
  children,
  defaultAppStoreValues
}) => {
  const updateAppStore = useAppStoreWriter()
  updateAppStore(defaultAppStoreValues || {})
  return <> {children} </>
}

export const NotFound = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  const { path = '/', pathParams = {}, defaultAppStoreValues, queryParams = {} } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  const history = createMemoryHistory({ initialEntries: [routePath] })

  return (
    <AppStoreProvider>
      <StringsContext.Provider value={strings}>
        <AppStoreInitializer defaultAppStoreValues={defaultAppStoreValues}>
          <Router history={history}>
            <ModalProvider>
              <RestfulProvider base="/">
                <Switch>
                  <Route exact path={path}>
                    {props.children}
                  </Route>
                  <Route>
                    <NotFound />
                  </Route>
                </Switch>
              </RestfulProvider>
            </ModalProvider>
          </Router>
        </AppStoreInitializer>
      </StringsContext.Provider>
    </AppStoreProvider>
  )
}
