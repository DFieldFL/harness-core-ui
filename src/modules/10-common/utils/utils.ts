import type { IconName } from '@wings-software/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}
