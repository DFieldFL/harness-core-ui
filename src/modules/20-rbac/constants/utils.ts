export enum RbacResourceGroupTypes {
  DYNAMIC_RESOURCE_SELECTOR = 'DynamicResourceSelector'
}
