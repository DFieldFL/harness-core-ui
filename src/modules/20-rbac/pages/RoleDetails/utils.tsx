import produce from 'immer'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@common/interfaces/ResourceInterface'
import type { Permission, PermissionResponse } from 'services/rbac'

export const getPermissionMap = (permissionList?: PermissionResponse[]): Map<ResourceType, Permission[]> => {
  const permissionsMap: Map<ResourceType, Permission[]> = new Map()
  permissionList?.map(response => {
    if (response.permission.resourceType) {
      if (permissionsMap.has(response.permission.resourceType as ResourceType)) {
        permissionsMap.get(response.permission.resourceType as ResourceType)?.push(response.permission)
      } else permissionsMap.set(response.permission.resourceType as ResourceType, [response.permission])
    }
  })
  return permissionsMap
}

export const onPermissionChange = (
  permission: string,
  isAdd: boolean,
  permissions: string[],
  setPermissions: (value: React.SetStateAction<string[]>) => void
): void => {
  if (isAdd) {
    if (
      permission === PermissionIdentifier.EDIT_DASHBOARD ||
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    ) {
      setPermissions(_permissions => [...permissions, permission, PermissionIdentifier.VIEW_DASHBOARD])
    } else {
      setPermissions(_permissions => [...permissions, permission])
    }
  } else if (
    !(
      permission === PermissionIdentifier.VIEW_DASHBOARD &&
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    )
  ) {
    setPermissions(_permissions =>
      produce(_permissions, draft => {
        draft?.splice(permissions.indexOf(permission), 1)
      })
    )
  }
}
