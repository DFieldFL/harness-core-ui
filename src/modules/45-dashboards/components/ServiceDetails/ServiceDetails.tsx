import React from 'react'
import { Layout } from '@wings-software/uicore'
import { Page } from '@common/exports'
import { ServiceDetailsHeader } from '@dashboards/components/ServiceDetails/ServiceDetailsHeader/ServiceDetailsHeader'
import { ActiveServiceInstances } from '@dashboards/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstances'

const ServiceDetails: React.FC = () => {
  return (
    <Page.Body>
      <Layout.Vertical>
        <ServiceDetailsHeader />
        <ActiveServiceInstances />
      </Layout.Vertical>
    </Page.Body>
  )
}

export default ServiceDetails
