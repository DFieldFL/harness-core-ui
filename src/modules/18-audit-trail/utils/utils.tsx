import React from 'react'
import { Menu } from '@blueprintjs/core'
import type { ItemRenderer } from '@blueprintjs/select'
import { Layout, Text, Avatar, Color, MultiSelectOption } from '@wings-software/uicore'
import { isEmail } from '@common/utils/Validation'

export interface UserItem extends MultiSelectOption {
  email?: string
}

export const UserItemRenderer: ItemRenderer<UserItem> = (item, { handleClick }) => (
  <Menu.Item
    key={item.value.toString()}
    text={
      <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Avatar name={item.label} email={item.email || item.value.toString()} size="small" hoverCard={false} />
        <Layout.Vertical padding={{ left: 'small' }}>
          <Text color={Color.BLACK}>{item.label}</Text>
          <Text color={Color.GREY_700}>{item.email || item.value}</Text>
        </Layout.Vertical>
      </Layout.Horizontal>
    }
    onClick={handleClick}
  />
)

export const UserTagRenderer = (item: UserItem, validate = false): React.ReactNode => (
  <Layout.Horizontal key={item.value.toString()} flex spacing="small">
    <Avatar name={item.label} email={item.value.toString()} size="xsmall" hoverCard={false} />
    <Text color={validate && !isEmail(item.value.toString().toLowerCase()) ? Color.RED_500 : Color.BLACK}>
      {item.label}
    </Text>
  </Layout.Horizontal>
)
