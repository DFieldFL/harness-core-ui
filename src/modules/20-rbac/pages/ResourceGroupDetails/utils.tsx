import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { ResourceSelector } from 'services/cd-ng'

export const getSelectedResourcesMap = (
  resourceSelectorList?: ResourceSelector[]
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()

  resourceSelectorList?.map(resourceSelector => {
    if (resourceSelector.type === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      map.set(resourceSelector.resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    else map.set(resourceSelector.resourceType, resourceSelector.identifiers)
  })

  return map
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelector[] => {
  const resourceSelectors: ResourceSelector[] = []
  map.forEach((value, key) => {
    if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      resourceSelectors.push({
        type: RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR,
        resourceType: key
      })
    else {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.STATIC_RESOURCE_SELECTOR,
        resourceType: key,
        identifiers: value
      })
    }
  })
  return resourceSelectors
}
