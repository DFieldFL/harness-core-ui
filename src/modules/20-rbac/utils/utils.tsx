import React, { ReactNode } from 'react'
import { Menu } from '@blueprintjs/core'
import type { ItemRenderer } from '@blueprintjs/select'
import {
  IconName,
  Layout,
  MultiSelectOption,
  Text,
  Avatar,
  Color,
  ModalErrorHandlerBinding,
  getErrorInfoFromErrorObject,
  SelectOption
} from '@wings-software/uicore'
import { defaultTo } from 'lodash-es'
import type { StringsMap } from 'stringTypes'
import type { AccessControlCheckError, RoleAssignmentMetadataDTO, UserMetadataDTO } from 'services/cd-ng'
import { Scope } from '@common/interfaces/SecretsInterface'
import type {
  Assignment,
  RoleOption,
  ResourceGroupOption
} from '@rbac/modals/RoleAssignmentModal/views/UserRoleAssigment'
import { isEmail } from '@common/utils/Validation'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import RBACTooltip from '@rbac/components/RBACTooltip/RBACTooltip'
import type { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@common/interfaces/ResourceInterface'
import type { FeatureRequest } from 'framework/featureStore/featureStoreUtil'
import type { PermissionsRequest } from '@rbac/hooks/usePermission'
import { FeatureWarningTooltip } from '@common/components/FeatureWarning/FeatureWarningWithTooltip'
import type { UseStringsReturn } from 'framework/strings'
import css from './utils.module.scss'

export interface UserItem extends MultiSelectOption {
  email?: string
}

export enum PrincipalType {
  USER = 'USER',
  USER_GROUP = 'USER_GROUP',
  SERVICE = 'SERVICE_ACCOUNT'
}

export const getRoleIcon = (roleIdentifier: string): IconName => {
  switch (roleIdentifier) {
    case '_account_viewer':
    case '_organization_viewer':
    case '_project_viewer':
      return 'viewerRole'
    case '_account_admin':
    case '_organization_admin':
    case '_project_admin':
      return 'adminRole'
    case '_pipeline_executor':
      return 'pipeline-executor'
    default:
      return 'customRole'
  }
}

export const UserTagRenderer = (item: UserItem, validate = false): React.ReactNode => (
  <Layout.Horizontal key={item.value.toString()} flex spacing="small">
    <Avatar name={item.label} email={item.value.toString()} size="xsmall" hoverCard={false} />
    <Text color={validate && !isEmail(item.value.toString().toLowerCase()) ? Color.RED_500 : Color.BLACK}>
      {item.label}
    </Text>
  </Layout.Horizontal>
)

export const UserItemRenderer: ItemRenderer<UserItem> = (item, { handleClick }) => (
  <Menu.Item
    key={item.value.toString()}
    text={
      <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Avatar name={item.label} email={item.email || item.value.toString()} size="small" hoverCard={false} />
        <Layout.Vertical padding={{ left: 'small' }}>
          <Text color={Color.BLACK}>{item.label}</Text>
          <Text color={Color.GREY_700}>{item.email || item.value}</Text>
        </Layout.Vertical>
      </Layout.Horizontal>
    }
    onClick={handleClick}
  />
)

export enum InvitationStatus {
  USER_INVITED_SUCCESSFULLY = 'USER_INVITED_SUCCESSFULLY',
  USER_ADDED_SUCCESSFULLY = 'USER_ADDED_SUCCESSFULLY',
  USER_ALREADY_ADDED = 'USER_ALREADY_ADDED',
  USER_ALREADY_INVITED = 'USER_ALREADY_INVITED',
  FAIL = 'FAIL'
}

interface HandleInvitationResponse {
  responseType: InvitationStatus
  getString: (key: keyof StringsMap, vars?: Record<string, any> | undefined) => string
  showSuccess: (message: string | ReactNode, timeout?: number, key?: string) => void
  modalErrorHandler?: ModalErrorHandlerBinding
  onSubmit?: () => void
  onUserAdded?: () => void
}

export const handleInvitationResponse = ({
  responseType,
  getString,
  showSuccess,
  modalErrorHandler,
  onSubmit,
  onUserAdded
}: HandleInvitationResponse): void => {
  switch (responseType) {
    case InvitationStatus.USER_INVITED_SUCCESSFULLY: {
      onSubmit?.()
      return showSuccess(getString('rbac.usersPage.invitationSuccess'))
    }
    case InvitationStatus.USER_ADDED_SUCCESSFULLY: {
      onUserAdded?.()
      return showSuccess(getString('rbac.usersPage.userAddedSuccess'))
    }
    case InvitationStatus.USER_ALREADY_ADDED:
      return showSuccess(getString('rbac.usersPage.userAlreadyAdded'))
    case InvitationStatus.USER_ALREADY_INVITED:
      return showSuccess(getString('rbac.usersPage.userAlreadyInvited'))
    default:
      return modalErrorHandler?.showDanger(getString('rbac.usersPage.invitationError'))
  }
}

export const getScopeBasedManagedResourceGroup = (
  scope: Scope,
  getString: UseStringsReturn['getString']
): SelectOption => {
  switch (scope) {
    case Scope.ACCOUNT:
      return {
        label: getString('rbac.allAccountResources'),
        value: '_all_account_level_resources'
      }
    case Scope.ORG:
      return {
        label: getString('rbac.allOrgResources'),
        value: '_all_organization_level_resources'
      }
    case Scope.PROJECT:
      return {
        label: getString('rbac.allProjectResources'),
        value: '_all_project_level_resources'
      }
    default:
      return {
        label: getString('rbac.allResources'),
        value: '_all_resources'
      }
  }
}

export const getScopeBasedDefaultAssignment = (
  scope: Scope,
  getString: UseStringsReturn['getString'],
  isCommunity: boolean
): Assignment[] => {
  if (isCommunity) {
    return []
  } else {
    const resourceGroup: ResourceGroupOption = {
      managedRoleAssignment: true,
      ...getScopeBasedManagedResourceGroup(scope, getString)
    }
    switch (scope) {
      case Scope.ACCOUNT:
        return [
          {
            role: {
              label: getString('common.accViewer'),
              value: '_account_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.ORG:
        return [
          {
            role: {
              label: getString('common.orgViewer'),
              value: '_organization_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.PROJECT:
        return [
          {
            role: {
              label: getString('common.projectViewer'),
              value: '_project_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      default:
        return []
    }
  }
}

export const isAssignmentFieldDisabled = (value: RoleOption | ResourceGroupOption): boolean => {
  if (value.assignmentIdentifier || value.managedRoleAssignment) {
    return true
  }
  return false
}
export const isDynamicResourceSelector = (value: string | string[]): boolean => {
  if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR) {
    return true
  }
  return false
}

interface ErrorHandlerProps {
  data: AccessControlCheckError
}

export const getRBACErrorMessage = (error: ErrorHandlerProps): string | React.ReactElement => {
  const err = error?.data
  if (err?.code === 'NG_ACCESS_DENIED' && err?.failedPermissionChecks?.length) {
    const { permission, resourceType, resourceScope } = err.failedPermissionChecks[0]
    if (permission && resourceType && resourceScope) {
      return (
        <RBACTooltip
          permission={permission as PermissionIdentifier}
          resourceType={resourceType as ResourceType}
          resourceScope={resourceScope}
          className={css.tooltip}
        />
      )
    }
  }
  return getErrorInfoFromErrorObject(error)
}

export const getAssignments = (roleBindings: RoleAssignmentMetadataDTO[]): Assignment[] => {
  return (
    roleBindings?.map(roleAssignment => {
      return {
        role: {
          label: roleAssignment.roleName,
          value: roleAssignment.roleIdentifier,
          managed: roleAssignment.managedRole,
          assignmentIdentifier: roleAssignment.identifier,
          managedRoleAssignment: roleAssignment.managedRoleAssignment
        },
        resourceGroup: {
          label: roleAssignment.resourceGroupName || '',
          value: roleAssignment.resourceGroupIdentifier || '',
          managedRoleAssignment: roleAssignment.managedRoleAssignment,
          assignmentIdentifier: roleAssignment.identifier
        }
      }
    }) || []
  )
}

interface FeatureProps {
  featureRequest: FeatureRequest
  isPermissionPrioritized?: boolean
}

interface TooltipProps {
  permissionRequest?: Omit<PermissionsRequest, 'permissions'> & { permission: PermissionIdentifier }
  featureProps?: FeatureProps
  canDoAction: boolean
  featureEnabled: boolean
}

interface TooltipReturn {
  tooltip?: React.ReactElement
}

export function getTooltip({
  permissionRequest,
  featureProps,
  canDoAction,
  featureEnabled
}: TooltipProps): TooltipReturn {
  // if permission check override the priorirty
  if (featureProps?.isPermissionPrioritized && permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  // feature check by default take priority
  if (featureProps?.featureRequest && !featureEnabled) {
    return {
      tooltip: <FeatureWarningTooltip featureName={featureProps?.featureRequest.featureName} />
    }
  }

  // permission check
  if (permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  return {}
}

export const getUserName = (user: UserMetadataDTO): string => {
  return defaultTo(user.name, user.email)
}
