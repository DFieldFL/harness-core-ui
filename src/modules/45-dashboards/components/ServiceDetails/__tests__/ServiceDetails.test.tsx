import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import ServiceDetails from '@dashboards/components/ServiceDetails/ServiceDetails'
import * as cdngServices from 'services/cd-ng'

jest.mock('highcharts-react-official', () => () => <></>)

jest.spyOn(cdngServices, 'useGetActiveServiceInstanceCountBreakdown').mockImplementation(() => {
  return { loading: false, error: false, data: [], refetch: jest.fn() } as any
})

jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
  return {
    mutate: () => Promise.resolve({ loading: true, data: [] })
  } as any
})

jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
  return { loading: false, error: false, data: [], refetch: jest.fn() } as any
})

describe('ServiceDetails', () => {
  test('should render ServiceDetails', () => {
    const { container } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <ServiceDetails />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
