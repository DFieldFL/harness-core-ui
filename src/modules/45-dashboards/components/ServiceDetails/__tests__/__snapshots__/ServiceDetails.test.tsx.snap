// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceDetails should render ServiceDetails 1`] = `
<div>
  <div
    class="Layout--vertical StyledProps--main"
  >
    <div
      class="Layout--horizontal StyledProps--main container standard StyledProps--flex StyledProps--padding-left-xlarge StyledProps--padding-right-xlarge"
    >
      <div
        class="Layout--vertical StyledProps--main"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
        >
          <div
            class="breadcrumbs"
          >
            <div>
              <a
                class="breadcrumb"
                href="/account/dummy/undefined/orgs/dummy/projects/dummy/dashboard"
              />
              <span
                class="bp3-icon StyledProps--main separator"
                data-icon="main-chevron-right"
              >
                <svg
                  height="11"
                  viewBox="0 0 8 14"
                  width="11"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                    fill="#4d5969"
                  />
                </svg>
              </span>
            </div>
            <div>
              <a
                class="breadcrumb"
                href="/account/dummy/undefined/orgs/dummy/projects/dummy/services"
              >
                services
              </a>
              <span
                class="bp3-icon StyledProps--main separator"
                data-icon="main-chevron-right"
              >
                <svg
                  height="11"
                  viewBox="0 0 8 14"
                  width="11"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                    fill="#4d5969"
                  />
                </svg>
              </span>
            </div>
            <div>
              <span
                class="lastBreadcrumb"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="StyledProps--font StyledProps--main pageBody"
      data-name="page-body"
    >
      <div
        class="Layout--horizontal StyledProps--main StyledProps--margin-left-xlarge StyledProps--margin-right-xlarge StyledProps--margin-top-large StyledProps--margin-bottom-large"
      >
        <div
          class="Layout--vertical StyledProps--main StyledProps--margin-right-xlarge"
        >
          <div
            class="bp3-card bp3-elevation-0 Card--card activeServiceInstances"
          >
            <div
              class="Layout--vertical StyledProps--main"
              style="height: 100%;"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-weight-bold StyledProps--color StyledProps--color-grey600 StyledProps--margin-bottom-medium"
              >
                dashboards.serviceDashboard.activeServiceInstancesLabel
              </p>
              <div
                class="Layout--horizontal StyledProps--main activeServiceInstancesHeader StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start StyledProps--padding-bottom-small"
              >
                <p
                  class="StyledProps--font StyledProps--main instanceCount StyledProps--font-weight-bold StyledProps--color StyledProps--color-grey800 StyledProps--margin-right-large"
                >
                  0
                </p>
                <div
                  class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
                  style="height: 100%;"
                >
                  <div
                    class="Layout--vertical StyledProps--main pieChartLabelContainerStyles StyledProps--flex StyledProps--flex-alignItems-flex-end StyledProps--padding-top-xsmall StyledProps--padding-bottom-xsmall StyledProps--padding-right-large"
                    style="height: 100%;"
                  >
                    <p
                      class="StyledProps--font StyledProps--main pieChartLabelStyles StyledProps--font-size-xsmall StyledProps--color StyledProps--color-grey500"
                    >
                      dashboards.serviceDashboard.nonProd (0)
                    </p>
                    <p
                      class="StyledProps--font StyledProps--main pieChartLabelStyles StyledProps--font-size-xsmall StyledProps--color StyledProps--color-grey500"
                    >
                      dashboards.serviceDashboard.prod (0)
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="Layout--vertical StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
                data-test="ActiveServiceInstancesEmpty"
                style="height: 100%;"
              >
                <img
                  height="100"
                  src="test-file-stub"
                  style="align-self: center;"
                  width="150"
                />
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400 StyledProps--margin-top-medium"
                >
                  dashboards.serviceDashboard.noDeployments
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Layout--vertical StyledProps--main fullWidth"
        >
          <div
            class="bp3-card bp3-elevation-0 Card--card card"
          >
            <div
              class="bp3-card bp3-elevation-0 Card--card card"
            >
              <div
                class="Layout--vertical StyledProps--main"
                style="height: 100%;"
              >
                <div
                  class="StyledProps--font StyledProps--main timeRange"
                >
                  <span
                    class="bp3-popover-wrapper timeRangeSelector"
                  >
                    <span
                      class="bp3-popover-target"
                    >
                      <button
                        class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main timeRangeButton StyledProps--font StyledProps--font-size-xsmall Button--with-current-color Button--without-shadow"
                        style="height: 20px;"
                        type="button"
                      >
                        <span
                          class="bp3-button-text"
                        >
                          dashboards.serviceDashboard.6months
                        </span>
                        <span
                          class="bp3-icon bp3-icon-caret-down StyledProps--main"
                          icon="caret-down"
                        >
                          <svg
                            data-icon="caret-down"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-down
                            </desc>
                            <path
                              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </div>
                <div
                  class="Layout--vertical StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
                  data-test="deploymentsWidgetEmpty"
                  style="height: 100%;"
                >
                  <img
                    height="100"
                    src="test-file-stub"
                    style="align-self: center;"
                    width="150"
                  />
                  <p
                    class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400 StyledProps--margin-top-medium"
                  >
                    dashboards.serviceDashboard.noDeployments
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
